### PR TITLE
development/runtime/testing

### PR DESCRIPTION
**Pull request by 4meta5**
*will align with recipes* - WIP

### CODE REVIEW

**Comment by shawntabrizi** [comment]
@4meta5 I started working on Errors in another PR, feel free to contribute there.

File: docs/development/module/tests.md [hunk 1]
```diff
@@ -8,22 +8,17 @@ Runtime tests allow you to verify the logic in your runtime module by mocking a
...
+```bash
+cargo test <optional: test_name>
 ```
 
 ## Mock Runtime Environment
 
 To support the specific needs of testing a Substrate runtime, you need to construct a "mock" runtime environment. This involves importing the logic of other runtime modules, such as the System module, and establishing runtime storage.
 
-In general, we suggest looking at and mimicking the various examples of runtime tests in the Substrate Runtime Module Library (SRML). This document won't be as comprehensive as what is provided in our SRML.
+In general, we suggest looking at and mimicking the various examples of runtime tests in `paint`. This document won't be as comprehensive as what is provided in `paint`.
```
Comment (JoshOrndorff) [inline docs/development/module/tests.md]: I think "paint" is outdated. "pallete" is in now.

**Comment by 4meta5** [comment]
I made the existing content more complete, but there is some redundancy with the [new chapter on testing](https://github.com/substrate-developer-hub/recipes/pull/87) in the recipes.

I see the following options:
1. wait until https://github.com/substrate-developer-hub/recipes/pull/87 is merged and then merge this (because the recipes is still heavily referenced in this doc)
2. throw this out and just have the recipes chapter on testing. It is my opinion that this topic is a bit more hands-on and aligns with the purpose of the recipes, which is to encourage tinkering with prepared examples. This decision to not put this in docs would be justified by the fact that the recipes covers it and this information is referenced from the devhub
3. duplicate content between https://github.com/substrate-developer-hub/recipes/pull/87 and this page

I'd rather not do (3). Once the recipes PR is merged, I'll onlt rebase this branch to catch up to master if there is interest in (1). Otherwise, I'm leaning towards (2) because of the reasons mentioned above. If there's silence, I'll close this.

File: docs/development/module/tests.md [hunk 2]
```diff
@@ -116,16 +128,6 @@ fn my_runtime_test() {
 }
 ```
 
-## Next Steps
```
Comment (shawntabrizi) [inline docs/development/module/tests.md]: Please follow the outline of the "next steps" laid out here.

This is one thing I will be firm on keeping consistant (and we have some CSS stuff which checks for these items on pages)

File: docs/development/module/tests.md [hunk 3]
```diff
@@ -6,104 +6,116 @@ Runtime tests allow you to verify the logic in your runtime module by mocking a
...
+	type Balance = u64;
+	type OnNewAccount = ();
+	type OnFreeBalanceZero = ();
+	type Event = ();
+	type TransferPayment = ();
+	type DustRemoval = ();
+	type ExistentialDeposit = ExistentialDeposit;
+	type TransferFee = TransferFee;
+	type CreationFee = CreationFee;
+}
```
Comment (shawntabrizi) [inline docs/development/module/tests.md]: I cant make a github suggestion here, but I think we should just keep this minimal:

```
impl balances::Trait for TestRuntime {
	type Balance = u64;
        //--snip//
}
```

All the other parts of the code seem irrelevant for the example.

File: docs/development/module/tests.md [hunk 4]
```diff
@@ -116,16 +128,6 @@ fn my_runtime_test() {
...
-
-### Examples
-
-TODO
-
-### References
+## References and Examples
 
-TODO
+The [testing](https://substrate.dev/recipes/base/testing/index.html) chapter of the [Substrate Recipes](https://github.com/substrate-developer-hub/recipes/) compliments the samples shown above, and the [kitchen](https://github.com/substrate-developer-hub/recipes/tree/master/kitchen) provides an environment to run the tests, change the logic, and tinker with the code.
```
Comment (shawntabrizi) [inline docs/development/module/tests.md]: Testing link is broken

**Comment by shawntabrizi** [comment]
I am not in favor of (2). I think the way you have set it up here is pretty good :)

I might look to add some more details and small things that people might expect to find in documentation, but definitely not looking to repeat all the content between the two.

I think if you fix the nits here, we can merge this as a good first step, and iterate as users ask us more questions.

File: docs/development/module/tests.md [hunk 5]
```diff
@@ -6,104 +6,116 @@ Runtime tests allow you to verify the logic in your runtime module by mocking a
...
-In general, we suggest looking at and mimicking the various examples of runtime tests in the Substrate Runtime Module Library (SRML). This document won't be as comprehensive as what is provided in our SRML.
+### Mock Runtime Storage
 
-### Mock Runtime Modules
+The [`runtime-io`](https://crates.parity.io/sr_io/index.html#enums) crate exposes a [`TestExternalities`](https://crates.parity.io/sr_io/type.TestExternalities.html) implementation frequently used for mocking storage in tests. It is the type alias for an in-memory, hashmap-based externalities implementation in [`substrate_state_machine`](https://crates.parity.io/substrate_state_machine/index.html)] referred to as [`TestExternalities`](https://crates.parity.io/substrate_state_machine/struct.TestExternalities.html).
 
-To test your module, you will need to construct a configuration type (`Test`) which implements each of the configuration traits of modules you want to use.
+In the [basic mock runtime's recipe](https://substrate.dev/recipes/testing/mock.html), an `ExtBuilder` object is defined to build an instance of [`TestExternalities`](https://crates.parity.io/sr_io/type.TestExternalities.html).
 
 ```rust
```
Comment (shawntabrizi) [inline docs/development/module/tests.md]: Why is this removed?

File: docs/development/module/tests.md [hunk 6]
```diff
@@ -6,104 +6,116 @@ Runtime tests allow you to verify the logic in your runtime module by mocking a
...
-		MyModule::on_finalize(System::block_number());
-		System::on_finalize(System::block_number());
-		System::set_block_number(System::block_number() + 1);
-		System::on_initialize(System::block_number());
-		MyModule::on_initialize(System::block_number());
+		let current_block = System::block_number();
+		ExampleModule::on_finalize(current_block);
+		System::on_finalize(current_block);
+		System::set_block_number(current_block + 1u64.into());
+		System::on_initialize(current_block);
```
Comment (shawntabrizi) [inline docs/development/module/tests.md]: This is wrong. You call `on_initialize` on a old block number...

The syntax I provide does not run into this error.

File: docs/development/module/tests.md (hunk 4 above)
Comment (4meta5) [reply docs/development/module/tests.md]: It won't work until the PR in recipes #87 is merged.

File: docs/development/module/tests.md (hunk 6 above)
Comment (4meta5) [reply docs/development/module/tests.md]: good catch, 
```suggestion
		System::on_initialize(current_block + 1.into());
```

Comment (shawntabrizi) [reply docs/development/module/tests.md]: 👎 You should use the old syntax and apply the change to both lines.

Comment (4meta5) [reply docs/development/module/tests.md]: To be clear, it is better to call `System::block_number()` four times in this method instead of calling it once and then using that value for each inner method call? 

Can you explain why the suggestion is wrong?

File: docs/development/module/tests.md (hunk 2 above)
Comment (4meta5) [reply docs/development/module/tests.md]: I had renamed it `References and Examples` but ok, I'll change it back

File: docs/development/module/tests.md (hunk 6 above)
Comment (4meta5) [reply docs/development/module/tests.md]: Done

File: docs/development/module/tests.md (hunk 5 above)
Comment (4meta5) [reply docs/development/module/tests.md]: added back, must have been a mistake

File: docs/development/module/tests.md (hunk 3 above)
Comment (4meta5) [reply docs/development/module/tests.md]: good call, recently addressed in last commit

File: docs/development/module/tests.md [hunk 7]
```diff
@@ -6,106 +6,113 @@ Runtime tests allow you to verify the logic in your runtime module by mocking a
...
 
-```rust
+```rust, ignore
 fn run_to_block(n: u64) {
 	while System::block_number() < n {
-		MyModule::on_finalize(System::block_number());
+		ExampleModule::on_finalize(System::block_number());
 		System::on_finalize(System::block_number());
-		System::set_block_number(System::block_number() + 1);
+		System::set_block_number(System::block_number());
```
Comment (shawntabrizi) [inline docs/development/module/tests.md]:
```suggestion
		System::set_block_number(System::block_number() + 1);
```

File: docs/development/module/tests.md (hunk 6 above)
Comment (JoshOrndorff) [reply docs/development/module/tests.md]: It seems like minimizing storage reads is a good idea how about something like this:

```rust
let mut current_block = System::block_number();
while current_block < n {
  // on finalize calls
  current_block += 1;
  System::set_block_number(current_block);
  // on initialize calls
}
```

File: docs/development/module/events.md [hunk 8]
```diff
@@ -105,11 +105,11 @@ The Substrate [JSON RPC](development/front-end/json-rpc.md) does not directly ex
 
 ### Examples
 
-View the following Substrate recipes to find examples of how runtime events are used:
+View the following substrate [recipes](https://github.com/substrate-developer-hub/recipes) to find examples of how runtime events are used:
```
Comment (JoshOrndorff) [inline docs/development/module/events.md]:
```suggestion
View the following Substrate [recipes](https://github.com/substrate-developer-hub/recipes) to find examples of how runtime events are used:
```

File: docs/development/module/events.md [hunk 9]
```diff
@@ -105,11 +105,11 @@ The Substrate [JSON RPC](development/front-end/json-rpc.md) does not directly ex
...
 
-* [A simple module which implements standard events]( TODO ).
+* [A module which implements standard events]( TODO )
```
Comment (JoshOrndorff) [inline docs/development/module/events.md]:
```suggestion
* [A pallet which implements standard events]( https://github.com/substrate-developer-hub/recipes/blob/master/kitchen/modules/last-caller/src/lib.rs)
```

File: docs/development/module/events.md [hunk 10]
```diff
@@ -105,11 +105,11 @@ The Substrate [JSON RPC](development/front-end/json-rpc.md) does not directly ex
...
 
-* [A module which does not emit events generic types]( TODO ).
+* [A module which does not emit events with generic types](https://github.com/substrate-developer-hub/recipes/blob/master/kitchen/adder/src/lib.rs)
```
Comment (JoshOrndorff) [inline docs/development/module/events.md]:
```suggestion
* [A pallet which does not emit events with generic types](https://github.com/substrate-developer-hub/recipes/blob/master/kitchen/adder/src/lib.rs)
```

File: docs/development/module/tests.md [hunk 11]
```diff
@@ -6,104 +6,116 @@ Runtime tests allow you to verify the logic in your runtime module by mocking a
...
-	});
-}
+```bash
+cargo test <optional: test_name>
 ```
 
 ## Mock Runtime Environment
 
-To support the specific needs of testing a Substrate runtime, you need to construct a "mock" runtime environment. This involves importing the logic of other runtime modules, such as the System module, and establishing runtime storage.
+To test a Substrate runtime, construct a mock runtime environment. The [substrate recipes](https://substrate.dev/recipes/testing/mock.html) provides a simple example of using a mocked runtime to test module logic. For more examples, see the `mock.rs` and `test.rs` files in most substrate modules.
```
Comment (JoshOrndorff) [inline docs/development/module/tests.md]:
```suggestion
To test a Substrate runtime, construct a mock runtime environment. The [Substrate recipes](https://substrate.dev/recipes/testing/mock.html) provides a simple example of using a mocked runtime to test module logic. For more examples, see the `mock.rs` and `test.rs` files in most substrate pallets.
```

File: docs/development/module/tests.md [hunk 12]
```diff
@@ -6,104 +6,116 @@ Runtime tests allow you to verify the logic in your runtime module by mocking a
...
 
-In general, we suggest looking at and mimicking the various examples of runtime tests in the Substrate Runtime Module Library (SRML). This document won't be as comprehensive as what is provided in our SRML.
+### Mock Runtime Storage
 
-### Mock Runtime Modules
+The [`runtime-io`](https://crates.parity.io/sr_io/index.html#enums) crate exposes a [`TestExternalities`](https://crates.parity.io/sr_io/type.TestExternalities.html) implementation frequently used for mocking storage in tests. It is the type alias for an in-memory, hashmap-based externalities implementation in [`substrate_state_machine`](https://crates.parity.io/substrate_state_machine/index.html)] referred to as [`TestExternalities`](https://crates.parity.io/substrate_state_machine/struct.TestExternalities.html).
 
-To test your module, you will need to construct a configuration type (`Test`) which implements each of the configuration traits of modules you want to use.
+In the [basic mock runtime's recipe](https://substrate.dev/recipes/testing/mock.html), an `ExtBuilder` object is defined to build an instance of [`TestExternalities`](https://crates.parity.io/sr_io/type.TestExternalities.html).
```
Comment (JoshOrndorff) [inline docs/development/module/tests.md]:
```suggestion
In the [basic mock runtime's recipe](https://substrate.dev/recipes/testing/mock.html), an `ExtBuilder` object is defined to build an instance of [`TestExternalities`](https://substrate.dev/rustdocs/master/sr_io/type.TestExternalities.html).
```

@shawntabrizi asked me the other day to start using the substrate.dev links for rustdocs.

File: docs/development/module/tests.md [hunk 13]
```diff
@@ -6,104 +6,116 @@ Runtime tests allow you to verify the logic in your runtime module by mocking a
...
+impl ExtBuilder {
+	pub fn build() -> runtime_io::TestExternalities {
+		let mut storage = system::GenesisConfig::default().build_storage::<TestRuntime>().unwrap();
+		runtime_io::TestExternalities::from(storage)
+	}
+}
+```
 
-When implementing the configuration traits of modules you want to use in your tests, you should choose primitives which makes your life easier. For example:
+To create the test environment in unit tests, the build method is called to generate a `TestExternalities` using the default genesis configuration. Then, [`with_externalities`](https://crates.parity.io/substrate_externalities/fn.with_externalities.html) provides the runtime environment in which we may call the module's methods to test that storage, events, and errors behave as expected.
```
Comment (JoshOrndorff) [inline docs/development/module/tests.md]:
```suggestion
To create the test environment in unit tests, the build method is called to generate a `TestExternalities` using the default genesis configuration. Then, [`with_externalities`](https://crates.parity.io/substrate_externalities/fn.with_externalities.html) provides the runtime environment in which we may call the pallet's methods to test that storage, events, and errors behave as expected.
```

File: docs/development/module/tests.md [hunk 14]
```diff
@@ -6,104 +6,116 @@ Runtime tests allow you to verify the logic in your runtime module by mocking a
...
+		System::on_finalize(current_block);
+		System::set_block_number(current_block + 1u64.into());
+		System::on_initialize(current_block);
+		ExampleModule::on_initialize(System::block_number());
 	}
 }
 ```
 
-You can then use this function in your unit tests:
+`on_finalize` and `on_initialize` are only called from `ExampleModule` if the module's trait implements the `sr_primitives::traits::{OnInitialize, OnFinalize}` traits to execute the logic encoded in the runtime methods before and after each block respectively.
```
Comment (JoshOrndorff) [inline docs/development/module/tests.md]: Module -> Pallet

File: docs/development/module/tests.md (hunk 12 above)
Comment (4meta5) [reply docs/development/module/tests.md]: good to know

File: docs/development/module/tests.md [hunk 15]
```diff
@@ -94,7 +94,7 @@ Account 1 has balance 10, account 2 has balance 20, and so on.
 
 It will be useful to simulate block production to verify that expected behavior holds during block time dependent changes. 
 
-A simple way of doing this increments the System module's block number between `on_initialize` and `on_finalize` calls from all modules with `System::block_number()` as the sole input.
+A simple way of doing this increments the System module's block number between `on_initialize` and `on_finalize` calls from all modules with `System::block_number()` as the sole input. While it is important for runtime code to [cache calls](https://substrate.dev/recipes/storage/cache.html) to storage or the system module, the test environment scaffolding should prioritize readability to facilitate future maintenance.
```
Comment (4meta5) [inline docs/development/module/tests.md]: The UI is making it unintuitive to reply to this https://github.com/substrate-developer-hub/substrate-developer-hub.github.io/pull/254#discussion_r352620401 so I'll just link it.

@JoshOrndorff It's a test so it doesn't really matter how many calls are made to system or storage. Computation/storage efficiency aren't that important in the test environment scaffolding. With that said, it is definitely not how things should be done for runtime development and I wonder if that might be confusing for readers of the doc. So I added this note.

File: docs/development/module/tests.md (hunk 14 above)
Comment (4meta5) [reply docs/development/module/tests.md]: addressed in most recent commit

File: docs/development/module/tests.md (hunk 2 above)
Comment (shawntabrizi) [reply docs/development/module/tests.md]: Sorry I was not clear enough. I mean the WHOLE format.

* Next Steps
   * Learn More
   * Examples
   * References

If you look at EVERY page it follows this pattern...

Comment (shawntabrizi) [reply docs/development/module/tests.md]: Notice too the action item are all bullets, so forming your actions as bullets too would be best.

File: docs/development/module/tests.md [hunk 16]
```diff
@@ -6,126 +6,123 @@ Runtime tests allow you to verify the logic in your runtime module by mocking a
...
 }
 ```
 
-You can then use this function in your unit tests:
+`on_finalize` and `on_initialize` are only called from `ExampleModule` if the pallet's trait implements the `sr_primitives::traits::{OnInitialize, OnFinalize}` traits to execute the logic encoded in the runtime methods before and after each block respectively.
+
+To use this function in unit tests,
 
-```rust
+```rust, ignore
```
Comment (shawntabrizi) [inline docs/development/module/tests.md]: Why are you adding `ignore` to this doc? It only has meaning in RustDocs right?

Comment (4meta5) [reply docs/development/module/tests.md]: deleted and IDK, it also has meaning in mdbook

**Comment by JoshOrndorff** [comment]
This look just about ready to go, and the recipes PR on testing is merged https://github.com/substrate-developer-hub/recipes/pull/87

Once @shawntabrizi 's comment about the Learn More section is addressed, I support merging.
